New module: Add module for manipulating the dconf database (system/dconf)

azaghal opened the pull request:
##### SUMMARY

A new module implementing ability to modify and retrieve values from the dconf database. This module can be useful when automating configuration of desktop environments such as GNOME or Cinnamon.

##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

dconf

##### ANSIBLE VERSION

```
ansible 2.4.0 (dconf-module c792ec282a) last updated 2017/03/27 23:32:48 (GMT +200)
  config file = 
  configured module search path = Default w/o overrides
  python version = 2.7.12 (default, Dec 13 2016, 14:29:11) [GCC 4.9.3]
```

##### ADDITIONAL INFORMATION

No additional information.

azaghal committed:
Implemented dconf module for manipulating and reading the dconf database.

gundalow commented:
Looks good. Just needs someone else to test and do a code review

azaghal commented:
Well, I just started _actually_ using the thing on remote host, and... The module will barf out becuase it can't access the user d-bus session (set via variable `DBUS_SESSION_BUS_ADDRESS` I believe). I'll try to explore a bit what's the best thing to do.

azaghal commented:
Found a simple solution, using `dbus-launch` as wrapper around `dconf`. I just realised I should probably add some info on dependencies (`dconf` tool + `dbus-launch` tool).

azaghal committed:
Fixed remote execution for the dconf module when state=present by wrapping the invocation with dbus-launch (dconf requires a running DBus user session). Updated documentation to mention external module dependencies.

azaghal committed:
Fixed remote execution for the dconf module when state=absent by wrapping the invocation with dbus-launch (dconf requires a running DBus user session).

azaghal commented:
Ok, added tiny fixes by using `dbus-launch` for `state=present` and `state=absent` (from what I can tell, read operation does not need this). Let me know if you prefer to have me squash the commits.

AsavarTzeth commented:
Using dbus-launch is a bad idea on systems that have dbus 1.8.0 and newer. Prefer dbus-run-session instead. If you use dbus-launch you need to run it in an unsafe shell and kill any processes it leaves behind, because dbus-launch won't kill them for you.

On other notes, not sure if this is completely duplicated effort, but here is my module for gsettings if you want inspiration: https://github.com/AsavarTzeth/ansible-gsetting

It offers compatibility with both newer and older dbus versions.

azaghal commented:
Well, that's a good heads-up, thanks. Couple of follow-up questions for you, though:

1. Given that you mention 1.8.0 or newer, I'm assuming that any older version of dbus does not have the `dbus-run-session` utility?

2. Could you tell me what would be the relation/difference between `gsettings` and `dconf` utility? Did a quick list of available entries, seems to be some overlap, but with different "paths". If the two provide same functionality (or one is super-set of the other), might be better to make a pick and improve that one instead.

3. I had a quick look at what you implemented, but I must say that I do not like the `use_unsafe_shell` option that much.

Although it would be more complicated, there could be two more commands run as a wrapper around the (in my case) `dconf` call that would call `dbus-launch`, then set environment variable for the `dconf` call, and finally kill the `dbus-launch` instance created. Although, the next thing that pops to my mind is race conditions in such a case... Plus it'd be a bit slower due to multiple subprocesses.

Otherwise something along the lines of `pipes.quote()` could help, although in Python 2.7 docs it seems to have a nice little red deprecation warning. Python 3 apparently has a `shlex.quote()`, but that's Python 3 :)

AsavarTzeth commented:
Answers in due order:

1. I believe your assumption is correct; that is what I gathered when I researched the topic. Although it does not hurt if someone else can verify it as well.

2. `gsettings` is a higher level tool for interacting with the `dconf` database. To be honest, at this point I think focusing on `dconf` is the right path, mainly because I don't think there is a way to do idempotent resets with `gsettings`. Looking at your code, it seems `dconf` is capable of this. The only downside with `dconf` is you need to reload the database if you want it to take effect in your current graphical session. If you are unlucky this may sometimes cause it to crash.

3. I too dislike the use of `use_unsafe_shell`, hate it even. I only provide it for backwards compatibility with Enterprise Linux systems. Plus it makes my code look a lot uglier, if one cares for such things.

I did experiment with alternative solutions, similar to what you propose, but never got it to work.

By the way this was my first python project, so have that in mind :) I am still very much a novice and my brain seems very inefficient when it comes to programming. So suffice to say, I am here to help, but guidance is very much appreciated.

mkoshelev commented:
If graphical session is already running, there's the `dbus-daemon` running too for the user in question. From my limited testing, if we supply the corresponding `DBUS_SESSION_BUS_ADDRESS` environment variable to dconf - the changes made via dconf seem to be reflected on the session immediately. If, on the other hand, bare `dbus-run-session` or `dbus-launch` is used to wrap dconf call while graphical session for the user is in progress, changes may be later rolled back to old values by graphical session.

For recent versions of dbus/systemd, `DBUS_SESSION_BUS_ADDRESS` is `unix:path=/run/user/<userid>/bus` so it may be worth to set `DBUS_SESSION_BUS_ADDRESS` in module for such cases. Using this path works both with and without running graphical session.

Unfortunately for older systems (f.e. RHEL7) it's not so simple to get path to dbus socket (and it only exists while graphical session is in progress). Setting `DISPLAY` seems to work instead, though we still need to get this value somehow (f.e. extract from `/proc/<pid>/environ` of existing dbus-launch/dbus-daemon processes for the user).

azaghal commented:
Hm... Some limited testing I've done (Debian 8 Stretch), changes seemed to be visible in the graphical session immediately - although I did not test what happens if you have the app using the setting running at the same time. Which could be a bit of a more complicated thing depending on the application at hand (how/if the app handles changes to configuration from the outside).

I was considering setting the environment variable as well (I think one quick-and-dirty solution I used was based on setting correct env variable on the task), but... Not sure how to handle all the different variations - as pointed out.

As for the listed path (`unix:path=/run/user/<userid>/bus`) - that is specific to systemd, I believe. On my Gentoo box, for example (using openrc), `DBUS_SESSION_BUS_ADDRESS` is unix:abstract=/tmp/dbus-XXXXX,guid=YYYYY`.

One thing I did notice is that remote logins via SSH would have the `DBUS_SESSION_BUS_ADDRESS` correctly set - but I guess that depending on both how Ansible handles environment, and whether sudo/su sanitise it along the way (if you use become_user), you may get different results. The `dbus-run-session` and `dbus-launch` look at least consistent in terms of being able to change the setting.

So, any suggestions what to progress with? I'm kind of keen on getting and maintaining this module in Ansible, but would find it useful to be able to somewhat wrap this pull request up at some point :)

The very least I'll try incorporating what @AsavarTzeth suggested for cleaning-up the `dbus-launch` remnants. I'm pondering on use of `dbus-run-session`, though, since `dbus-launch` works for both older and newer versions of D-Bus. Then again, if checking for its availability is cheap, shouldn't be much of a hassle. :)

mkoshelev commented:
I would say let's try to do the right thing (use existing dbus-daemon) when possible with fallback to  dbus-launch:
- check if `DBUS_SESSION_BUS_ADDRESS` is present in environment
- if not, try to extract it from this user's `dbus-daemon` processes (if present)
- failing all above, wrap dconf in `dbus-launch` and kill the corresponding dbus-daemon afterwards

I cobbled up a function to search for DBUS_SESSION_BUS_ADDRESS in running dbus-daemon environments. Tested to work on both RHEL7 and Fedora 25, please feel free to include if you like it.

```python
def find_dbus_path():
    """
    Search environments of user's dbus-daemon processes for
    DBUS_SESSION_BUS_ADDRESS environment variable

    Returns first found DBUS_SESSION_BUS_ADDRESS value or None
    """

    try:
        pids = [ p for p in os.listdir('/proc') if p.isdigit() ]
        for pid in pids:
            # search only this user's processes
            if os.stat("/proc/%s/environ" % pid).st_uid == os.getuid():
                with open("/proc/%s/cmdline" % pid) as FH_cmd:
                    # look only in dbus-daemon processes' environments
                    if FH_cmd.read().find('dbus-daemon') != -1:
                        with open( "/proc/%s/environ" % pid ) as FH_env:
                            for s in FH_env.read().split('\x00'):
                                if s.startswith('DBUS_SESSION_BUS_ADDRESS='):
                                    return s.replace('DBUS_SESSION_BUS_ADDRESS=', '', 1)
    except (OSError,IOError):
        return None
    return None
```

AsavarTzeth commented:
I have to admit I have been focusing too much on all the cases where the dbus-daemon does not exist. Any potential issues with never checking for an existing session were forgotten about in the process.

So I my thinking would probably mostly be inline with that of @mkoshelev. Although I would not dismiss `dbus-run-session` though. Quotes from `dbus-launch(1)`:

> To start a D-Bus session within a text-mode session, do not use dbus-launch. Instead, see dbus-run-session(1).

The [Debian mailing list](https://lists.debian.org/debian-devel/2016/08/msg00554.html) happen to contain an excellent page with reasons for preferring `dbus-run-session`, while keeping it optional (i.e. fallback in our case). They cover it way in way more detail than I ever could.

azaghal commented:
Great, thanks for all the feedback. I'll try to integrate the two approaches, and read through the Debian description :)

Might try getting some feedback on the IRC dev channel on all the subprocess'ing within the module as well - i.e. if there are any no-no's etc.

azaghal commented:
Well, finally got around to working on this again, and ran into issue when testing the  code that checks for existing sessions. Here is the scenario:

1. Make sure you are not logged-in into the machine in any way.

2. Log-in via SSH.

3. Verify that no D-Bus daemon is running.

4. Make sure the `DBUS_SESSION_BUS_ADDRESS` variable is not set:

        unset DBUS_SESSION_BUS_ADDRESS

5. Start D-Bus daemon via `dbus-launch`:

        dbus-launch

6. Take not of the `dbus-daemon` PID, then extract environment variables of the running process:

        strings /proc/PID/environ

7. At this point, at least on Debian Jessie/Stretch, you won't see the `DBUS_SESSION_BUS_ADDRESS` set for the process. So go figure... You have a running D-Bus daemon, but no reliable way to get the session address for it...

Based on what I can tell, though, if you have logged-in via graphical display manager beforehand, the session `dbus-daemon` will have this environment variable set - quite probably whatever starts the daemon (`systemd` I guess) sets this variable beforehand.

Unless someone comes-up with a new idea, I'll proceed with the implementation, knowing that detection is not 100% reliable (and making note of it in docs).

azaghal commented:
Actually, another option is to fail in case dbus-daemon is detected, but session address cannot be extracted...

@AsavarTzeth @mkoshelev Any preferences on doing it this way? Could maybe even add an extra parameter for forced behaviour, and do some warning logging if the parameter is set to `true`.

azaghal committed:
Updated dconf module implementation to make it more robust:

- Detect running D-Bus session, and reuse that one if possible.
- If detection fails, try launching process via dbus-run-session to avoid
  left-over D-Bus processes.
- As last resort run dbus-launch, and clean-up after the changes have been
  made.
- Updated documentation to mention new dependencies and to be more explicit
  about module limitations.

azaghal commented:
There we go, new implementation. Will wait for feedback.

AsavarTzeth commented:
Hold on I am writing a response at this moment.

AsavarTzeth commented:
I followed your steps to reproduce the supposed issue and I am not surprised at all. If you don't export DBUS_SESSION_BUS_ADDRESS then it won't be there. See the man page:

> With no arguments, dbus-launch will launch a session bus instance and print the address and PID of that instance to standard output.

If you want the PID to be set to DBUS_SESSION_BUS_ADDRESS you must do like I did in my module:

    export `dbus-launch`
    # alternatively:
    export $(dbus-launch)

But then you must use the unsafe shell option you did not like.

azaghal committed:
Fixed PEP8 errors reported by ansibot in dconf module.

azaghal commented:
Actually, what I was hoping for is that the `dbus-daemon` might change its own environment (probably does not make much sense, though ;). Now that I think a bit about it, some logic change might be needed for detecting existing session path.

My decision to read environment from running `dbus-daemon` might be completely wrong (in spite of working on Debian), and I should go through all the processes owned by the user + check if `dbus-daemon` is running at all. However, that still does not guarantee that the bus session path picked up is the one in use by running/existing `dbus-daemon` :/ So maybe I'd need to check every single path to see if I can connect to it before making a decision... I'll wait for the feedback though. :)

As a side-note, pushed small PEP8 fix while at it. If anyone would like some commit squashing, let me know.

AsavarTzeth commented:
Also to clarify things, (I am a bit tired at the moment). You currently need the missing variable `DBUS_SESSION_BUS_ADDRESS` to find existing sessions? And this is not available if you don't have a graphical session running?

AsavarTzeth commented:
Nope, I am way to tired to make sense of this :P. I am likely not much help in this matter anyhow. All I keep seeing is problems that were fixed with `dbus-run-session`.

I might try to make sense of it in the morning.

azaghal commented:
Maybe you can clarify the question, since I'm not 100% sure what you're asking me (more that I think there might be some context missing :).

In order to reuse existing D-Bus session for `dconf`, you need to set `DBUS_SESSION_BUS_ADDRESS` environment variable for it. And in order to extract it, you need to traverse the process tree, and get it out of one of the running processes.

Depending on how you login (GUI vs SSH vs CLI) and where (different distros with different init systems), you may have more or less luck locating it in this manner. In my current implementation, I actually take it out of the dbus-daemon's process. But I think that one is set for it by systemd on Debian, and might not be the same elsewhere.

But then, when you do locate a process where this variable is set, and pick it up, you're not quite sure if it is valid or not (i.e. the `dbus-daemon` that used that path might be dead).

AsavarTzeth commented:
Yes it seems it took me a while to understand why you need `DBUS_SESSION_BUS_ADDRESS`. Its moments like this I feel broken or stupid... Thank you for helping with that clarification. I should just have asked for that to begin with.

azaghal commented:
Nah, I think it's just that the whole thing is a bit more convoluted than it should be. Truth be told, I'm somewhat surprised how unreliable this thing with user sessions is.

AsavarTzeth commented:
I could help testing different distributions if you would like. That seems to be something I can manage even when my brain feels less functional.

AsavarTzeth commented:
Well I did some testing. I found another related issue; the dependency on psutil. While I am not too fond of it, since that means I need an addition installation task, that is not the main issue.

You will need to try the import and fail appropriately. Handle the failure in main().

    try:
        import psutil
        HAS_PSUTIL=True
    except:
        HAS_PSUTIL=False

Although to be honest, while the function by @mkoshelev is larger it does have less dependencies.

azaghal commented:
Good point, I will look into at least checking existence of `psutil`.

I decided to go with `psutil` as dependency to avoid writing code for killing process etc (and library feels pretty "Pythonic" to me).

If you are still up for testing how the environment variables are set on different distros, you could help me with doing something like:

1. Log-in via GUI and/or CLI.

2. Check if D-Bus session variable is set.

3. Run Python code similar to the one in `DBusWrapper._get_existing_dbus_session()` method (probably just printing to screen the variable).

I am thinking of changing the code to actually go through all user's processes, check for existence of environment variable, and then also try to connect to D-Bus using discovered session to ensure it is valid. Something like this could help (address being replaced with whatever got found):

```
dbus-send --address="unix:path=/run/user/1000/bus" --type=signal / com.example.signal_name
```

I'm still finding it funny how much code has to be written for all this, though :)

azaghal commented:
Oh, right, `psutil` is cross-platform as well (supports *BSDs). Not sure how much going through process tree can differ between, say, *BSD and GNU/Linux.

azaghal committed:
Updated dconf module implementation:

- Fail early if psutil library is not available on the system.
- Go through all of user's processes to locate a running D-Bus daemon.
- Test potential D-Bus session bus address before deciding to (re)use it.
- Added a couple of debug statements.
- Updated documentation to include dbus-send as requirement.

azaghal commented:
Here is another change. Main thing is to go through all of user's processes, and try to connect before declaring the existing D-Bus session bus address as usable. Also applied to check for module existence.

AsavarTzeth commented:
Well I finished doing some testing. I was going to try out more distributions and init systems but ran out of time. I used this [script](https://gist.github.com/AsavarTzeth/15f85c09d35ce6fa0c6f6135fbf52b89).

The results:

**CentOS 7**

- Init: systemd
- DE: gnome-shell
- DM: gdm
- Python: 2.7.5

**Debian 8**
- Init: systemd
- DE: gnome-shell
- DM: gdm
- Python: 2.7.9 and 3.4.2

Both got exceptions, stacktrace:

    File "./dconf.py", line 45, in <module>
      print (_get_existing_dbus_session())
    File "./dconf.py", line 26 in _get_existing_dbus_session
      if process_real_uid == uid and 'DBUS_SESSION_BUS_ADDRESS' in process.environ():
    AttributeError 'Process object has no attribute 'environ'

**Ubuntu 14.04.5**

- Init: upstart
- DE: unity
- DM: lightdm
- Python: 2.7.6 and 3.4.0

Got exception:

    File "./dconf.py", line 45, in <module>
      print (_get_existing_dbus_session())
    File "./dconf.py", line 22 in _get_existing_dbus_session
      for pid in psutil.pids():
    AttributeError: 'module' object has no attribute 'pids'

**Arch Linux**

- CLI: Success
- GUI: Success
- Init: systemd
- DE: gnome-shell
- DM: gdm
- Python: 2.7.13 and 3.6.1

**Fedora 25**

- CLI: Success
- GUI: Success
- Init: systemd
- DE: gnome-shell
- DM: gdm
- Python: 2.7.13

**Void Linux**

- CLI: Failed
- GUI: Success
- Init: runit
- DE: xfce4
- DM: lightdm
- Python: 2.7.13 and 3.5.3

azaghal commented:
First of all, thanks for super-extensive testing you've done :)

When it comes down to your summary, I would say that's pretty much in line with my experiences. I.e. if `systemd` is used, the environment variable will be available, with some other process managers it might not. Another thing I _think_ happens is that the environment variable won't be available when Ansible switches via `sudo` (I think I did some testing using the command module with `printenv` binary to test this).

I must say that I am a bit torn between keeping `psutil` and replacing it with custom code. On one hand, I doubt I could write better code than what psutil provides (plus it's quite convenient), on the other some of the distros will be maintained for a long time...

What versions of `psutil` did `CentOS 7` and `Ubuntu 14.04.5` ship with (and did you install them in both cases from standard repos)?

AsavarTzeth commented:
I think the Ansible become feature uses a login shell, although it was some time since I read the source. In any case, you are correct, using something like `sudo -i -u username` will not inherit `DBUS_SESSION_BUS_ADDRESS`.

This is a good thing, because sharing dbus session bus instances between users is not supported according to the documentation. It makes a lot of sense.

Now because it is not a proper login session (as far as systemd-logind is aware), a new dbus session won't be started. So yes, when we use become we must use the wrapper, no matter what distribution.

If you want something akin to changing user with sudo/su, including a full systemd user session, `macinectl shell <username>@` is one option. But that does not exist as a `become_method` in Ansible.

Versions of `psutil`:
1. `Ubuntu 14.04`: 1.2.1
2. `CentOS 7`: 2.2.1

In both cases I used the official repositories.

I can see why you are torn. It seems like a really annoying issue to solve... It really is that much harder when you try to make it perfect for everyone.

azaghal commented:
I'll try to discuss some things around the module's design on Thursday 2017-05-11 in a community meeting before proceeding further. This is actually more related to spawning of D-Bus daemon than the `psutil` dependency, though. Truth be told, I'm considering making a hard requirement on up-to-date `psutil` library to make life (and future maintenance) easier.

AsavarTzeth commented:
That sounds like a good idea. I think more points of view and more feedback would help a lot.

If you do decide on that hard requirement I guess I would be okay with that. Since my own use case is currently limited to Arch Linux I can get it all from the repositories.

abadger commented:
Sorry I missed the meeting today.  Afterwards, @bcoca and I had a long talk about it in #ansible-devel.  We had one central question which revolves around this bullet point from the Meeting comment:

> * In order to ensure integrity and avoid accidental overwrites of settings, one should not run multiple D-Bus user sessions.

Can you tell us more about that?  It looks like it's not a limitation of dbus itself (the docs on freedesktop.org appear to encourage use of dbus-run-session to launch programs with their own dbus daemons from shell scripts.) Is it a problem with dconf not locking its persistent backend store appropriately?

azaghal commented:
This issue in particular was mentioned in [one of the comments above](https://github.com/ansible/ansible/pull/23015#issuecomment-296398862). I have not tried to verify the findings, although on a quick glance I should be able to do so by executing `/usr/libexec/dconf-service` under two separate D-Bus user sessions (once I do, I'll try to get back to you).

Probably will need to test with some live app, since `dconf read` reads values directly (that one never needed a running D-Bus daemon).

azaghal commented:
Did some quick testing just now with keyboard layouts under Cinnamon, on Debian Stretch. Should be quite similar under Gnome as well, I believe. Keep in mind that `dbus-run-session` is available only on systems with D-Bus 1.8.0 and above.

Steps performed:

1. Log-in via GUI, and configure `us` keyboard layout as the only available layout.

2. Verify that keyboard layout indicator is not shown in the taskbar.

3. Launch terminal inside of GUI.

4. Change keyboard layout:

        dconf write /org/gnome/libgnomekbd/keyboard/layouts "['us', 'rs']"

5. Verify that keyboard layout indicator in taskbar is now updated, and you can pick between `English (US)` and `Serbian`.

6. Change keyboard layout again, this time using dedicated D-Bus session:

        dbus-run-session dconf write /org/gnome/libgnomekbd/keyboard/layouts "['us']"

7. Observe that the keyboard layout indicator has not changed, and that you can still pick between the two layouts previously configured.

8. Open keyboard layout configuration in the GUI. You will notice that the list of layouts includes just `English (US)`.

azaghal commented:
Now, here is another thing I noticed - if you try to run `dconf-service` twice, using same D-Bus session address, it will refuse to start the second time. This is quite probably (might be using wrong terminology here :) due to the fact that previous one is already listening on D-Bus on same path/directory (or whatever it's called).

So, it could be that `dconf-service` uses D-Bus for "locking" access (running the thing through `strace -f` to figure out if it opens any other file did not yield much success).

I hope this makes sense?

azaghal committed:
Updated dconf module implementation:

- Simplified module, removing all code for handling dbus-daemon, as discussed in
  a community meeting.
- Module user must ensure that D-Bus user session is available and specified
  either via module parameter or environment variable.
- Updated documentation for the change.

azaghal commented:
@AsavarTzeth After some discussions yesterday, decision was made not to manage the `dbus-daemon` from within the `dconf` module at all. Instead of this I've made changes to available parameters (one can specify the D-Bus user session address via module parameter or environment variable), and updated documentation accordingly.

Since I know you invested time in testing all of this, and I ended-up removing relevant functionality, I would like to extend my gratitude for the effort nevertheless - it is really appreciated :)

AsavarTzeth commented:
Yes I was afraid this would happen, but I can definitely see the reasoning behind it. I would have preferred just dropping the code that tries to find an existing session and replacing that with the mentioned module parameter.

Instead this means it is unusable to me, and I will be forced to maintain my own.

Examples of use cases that will not work:
- Using it within a chroot
- Using it with `become` and a user that is not already logged in

Good news is I can take what I have learnt here and fix the issues I had.
1. First I will drop the ugly `dbus-launch` legacy support in my own module, since I personally don't need it; I only added it in case I would upstream it (not relevant anymore)
2. I will probably try to rework it so it too uses `dconf` directly instead of `gsettings`.

Oh and you are very much welcome :) I will look forward to try the module again, nevertheless.

azaghal commented:
@AsavarTzeth Well, I did add a module option to specify the address? Or did you have something else in mind in terms of parameters?

AsavarTzeth commented:
@azaghal Specifying the address does not help if the `dbus-daemon` is not running in the first place. Until Ansible has a `become_method` that ensures dbus is available this simply is not possible.

What you mention would only work if the user has already logged in, either graphically, or in a shell on systems with newer systemd/dbus. We covered this already earlier.

azaghal commented:
Ok, just wanted to make sure I understood :)

As a side-note, at least on systemd under Debian Stretch the D-Bus user session address is predicatable (`unix:path=/run/user/UID/bus`) and auto-activated by systemd. Just something to keep in mind.

There have been some discussions on making a separate module for managing user's D-Bus session, for what it's worth.

AsavarTzeth commented:
I know the path is predictable, it still won't help. If a user is not logged in "`UID`" in `unix:path=/run/user/UID/bus`. It will auto activate if the user logs in, 

Unless you log in for real it will not be "auto-activated by systemd", using `dbus.socket` and/or `dbus.service`. Using tools like `sudo` or `su` to "log in" will not do this and therefore neither will `become`.

azaghal commented:
Ah... I just realised that the path will not be there at all as you have stated :/ Must have been testing on a system where I was already logged-in. Well, there goes that plan. I guess someone (or me) would have to look into yet another module (tm) implementation or keep a forked version -.-

The separate module was discussed [in community meeting, see around 15:14:47](https://meetbot.fedoraproject.org/ansible-meeting/2017-05-11/ansible_core_meeting.2017-05-11-15.00.log.html)

azaghal commented:
And now my docs for this module are incorrect too :P

azaghal commented:
I have given this a bit more thought, and I must say that in current (reduced) form the module is essentially useless to me as well. Even if a dedicated module is introduced for spawning/stopping D-Bus user session, this would mean wrapping calls to `dconf` with two more tasks, and effectively having Ansible playbooks report changes have been made on every run.

The only thing that comes to my mind is to:

1. Use environment variable for D-Bus user session (if set).
2. If not, traverse PID tree to locate D-Bus user session address.
3. If not, run `dconf` via `dbus-run-session`. This would at least kill the daemon as soon as `dconf` is done running.
4. If not (`dbus-run-session` not available because D-Bus is <1.8.0), fail the module.

Otherwise I think I'll just close the pull request instead (since I would need to maintain my own module for the above otherwise).

abadger commented:
<nod> In discussion my discussion with @bcoca I know that the problem of corruption of the backend store and the problem of a user's session being crashed were the main sticking points.  If we don't have those problems, only the problem that the settings may be rolled back if the user has a session open or that the settings may not take effect until the user logs out and back in then starting and stopping an extra d-bus user session might become more palatable to everyone.

@azaghal it seems from your testing and the comment you linked to that overwriting and not being in sync are the only problems expressed.  If so, I think we should put this back on the meeting agenda as new information/clarifying of existing information.

azaghal commented:
@abadger Ok, I'll add comment for another discussion on this one then. So much for the easy module to implement :)

azaghal commented:
@AsavarTzeth A bit late reply, but essentially I will be adding back some functionality related to D-Bus into the module. Namely, module will try to locate running D-Bus session, and if that does not exist, it'll run command via dbus-run-session (so only D-Bus 1.8.0+) - since  that is much safer than dbus-launch and you don't end-up managing daemon processes as such.

Before I go ahead, would this work for your use-case too? Since at the end module should be useful to people :)

azaghal committed:
Updated dconf module implementation:

- Add back ability to detect running D-Bus user session.
- Fail-back to using dbus-run-session if running session could not be detected.

azaghal committed:
PEP8 fix for dconf module.

AsavarTzeth commented:
@azaghal That would work for me. All distributions I use for desktop purposes has `dbus-run-session` available.

Although I am not as happy about `psutil` being re-added. I guess it is to support the distributions that still use /tmp and a randomized name for the bus address. Anyhow, I can live with it.

I do have an idea, where we keep the `dbus_address` option as a complementary feature. It should be possible to make the `_get_existing_dbus_session` function and `psutil` optional (but used by default), if the `dbus_address` option is used.

Since most major distributions use the deterministic path `unix:path=/run/user/UID/bus` and I could then happily manage it at the task level and in exchange I would not have to install `psutil`.

I don't ask for this now though. I say, continue as you planned and get the core feature set stable so we can get something usable. When you need testing to be done, just tell me and I will test as widely as possible, within the scope of the new requirements.

azaghal commented:
Ah... Well, yes, the `psutil` and specifically going through the process tree is meant to be used as a way to locate user sessions on distros that do not have standard path template (e.g. most that actually do not run `systemd`, I would say). And it should cover those with standard path template too.

The module is ready for testing, and I agree that the `dbus_address` could be reintroduced. I'll try to ask maybe on IRC or user list to see if anyone else is interested in the module.

P.S.
I hope this gets merged soon, this is starting to look like one of "those" threads :)

AsavarTzeth commented:
Alright, I will do some testing then. If it works well I will try do do a review and we can focus on getting this merged.

Good to know you agree with me on future additions.

AsavarTzeth commented:
Testing the module on Debian 9/Testing, Fedora 25, Arch Linux, OpenSUSE Tumbleweed, Ubuntu 17.04 and last Void Linux (for non-systemd), resulted in everything working as expected.

Methods of testing:
- Connecting to host via Ansible ssh connection
- First using ssh to connect to host and then running Ansible with local connection
- In a graphical session using Ansible with local connection
- Connecting to host via Ansible ssh connection and using become on another user (tested both logged in and not)

In all instances the changes were applied immediately on any user that is logged in and has an existing dbus session.

AsavarTzeth reviewed (changes requested):
Fix examples:
The current dconf path for changing keyboard layouts in Gnome is `/org/gnome/desktop/input-sources/sources` with a key being a list of tuples like `"[('xkb', 'us'), ('xkb', 'se')]"`

Cinnamon is a fork of older Gnome so it would naturally have fallen behind or chosen not to follow.

Fix typo:
The last changes brought a typo "Rests value for the specified key". Fix is obvious.

Suggestion:
Any reason why you cannot or will not use `module.get_bin_path('dbus-run-session')` from ansible.module_utils.basic? (line 155)

Also, I might have misunderstood how the suggest changes feature works. So I guess for all intents and purposes this is just a comment with suggestions.

azaghal committed:
Updated dconf module implementation:

- Introduce correct examples for Gnome DE.
- Rename existing examples to mark them as Cinnamon-specific.
- Use self.module.get_bin_path instead of custom check for dbus-run-session.
- Fixed typo in method documentation for DconfPreference.reset().

azaghal commented:
Updated everything according to comments. Examples got expanded to cover both Cinnamon and Gnome. Take note I have not tried this out against Gnome :) The module.get_bin_path thing is that I simply forgot about it (actually, got that one mentioned in another pull request review already recently - didn't know about it at all back then).

Not sure what you meant by the "there are not changes to see" comment, maybe you can clarify? Essentially, in `write` and `reset`, if value is the same when we read it, we return `False` in either case (we don't need to change value, and this would be true for check mode too). However, if the value differs, we fall into the `elif self.check_mode` to see if user does not want to apply changes, and merely wants to see if values would've changed. So we bail out there immediately if we are in check mode, and tell the user things "changed" with `return True` (but, since we are in check mode, this is actually just indicator that nothing needs to be changed).

AsavarTzeth commented:
@azaghal Regarding the "there are no changes to see" comment.

That was just me being confused by the GitHub review interface. This is the first time I used that feature. In my haste I obviously forgot to clarify that last comment targeted at GitHub not the module.

You see when I used this **GitHub** feature adds a "View Changes" button, even if I don't have write access. I kind of panicked when I noticed this and felt I had to explain why no changes can be viewed. Obviously I failed miserably.

AsavarTzeth reviewed (approved):
Looks pretty good now. I approve.

azaghal commented:
Ah, yes, I get confused trying to locate reviews other people have created too, truth be told :) There's something a bit counter-intuitive with review comments being mixed-in with general comments etc.

abadger commented:
Merged to devel.  thanks everyone for code and review and testing on various distros!

AsavarTzeth commented:
@jlacroix82 

That is `ansible_connection: local`.

You are currently trying to connect over ssh and it is either not running (most likely) or it is blocked in a firewall.

jlacroix82 commented:
> That is ansible_connection: local.

> You are most likely connecting over ssh and it is either not running (most likely) or it is blocked by a firewall.

I tried ansible_connection: local, and no difference. SSH is working fine on this machine. Also, all other tasks in this playbook are completing successfully, only the dconf module fails.

azaghal commented:
Not sure if separate ticket should be opened for this, but in either case - could you provide full minimal example that fails? I.e. hosts file + playbook + command run.

jlacroix82 commented:
Sure.

This is the playbook that fails: https://pastebin.com/Q8fFRZpF

Hosts file: https://pastebin.com/eQ3W1HAX

host_var: https://pastebin.com/Tbeh9kK3

jlacroix82 commented:
I have a theory that this is failing because it's being run as a user other than the user I'm logged in as. I have a user, cleverly named "ansible", that runs my playbooks. If I run dconf commands manually within bash, it fails unless I run the command as my user account.

azaghal commented:
Well, I did ask for a full minimal example to reproduce the issue :)

You will need to use `become` + `become_user` to switch to correct user whose settings you are modifying.

The Ansible dconf module does not require running X session. It does need D-Bus, and it does a couple of things to get its hands on it.

Could I ask you to open-up a separate ticket for this, since the pull request has already been merged? If you could describe the set-up in more detail and with concrete things I can just run to reproduce the issue, it could help me in debugging/understanding the issue :)

jlacroix82 commented:
Thanks for the response. It was become_user I was missing. I didn't include it because of a comment earlier in this thread that mentioned it was changed for every logged in user with a dbus session.

Now everything is working. I have noticed that keyboard shortcuts set by this module don't work (the keys and values are set properly, but they don't function in MATE at all) but I'd bet that's probably an issue with MATE and not this module.

Cheers! This module is awesome and you've done a great job.

azaghal commented:
Hm... That last issue (about shortcuts not working) could be two possible issues - one is wrong syntax for setting values, the other is that there's desyncronisation between the running session and `dconf` module (which happens due to limitations of detecting running D-Bus session etc). I _think_ - could be something third as well. I'll try to leave this comment thread die out, feel free to ping me on `#ansible` IRC channel in case you run into more issues, would be more than happy to help :)

jlacroix82 commented:
All the previous issues I had have been resolved. However, I am getting a new error since trying it on Ubuntu GNOME 16.04.2 today:

An exception occurred during task execution. To see the full traceback, use -vvv. The error was: AttributeError: 'Process' object has no attribute 'environ'
fatal: [phantom]: FAILED! => {"changed": false, "failed": true, "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_56yfEr/ansible_module_dconf.py\", line 384, in <module>\n    main()\n  File \"/tmp/ansible_56yfEr/ansible_module_dconf.py\", line 377, in main\n    changed = dconf.write(module.params['key'], module.params['value'])\n  File \"/tmp/ansible_56yfEr/ansible_module_dconf.py\", line 305, in write\n    dbus_wrapper = DBusWrapper(self.module)\n  File \"/tmp/ansible_56yfEr/ansible_module_dconf.py\", line 168, in __init__\n    self.dbus_session_bus_address = self._get_existing_dbus_session()\n  File \"/tmp/ansible_56yfEr/ansible_module_dconf.py\", line 194, in _get_existing_dbus_session\n    if process_real_uid == uid and 'DBUS_SESSION_BUS_ADDRESS' in process.environ():\nAttributeError: 'Process' object has no attribute 'environ'\n", "module_stdout": "", "msg": "MODULE FAILURE", "parsed": false}

AsavarTzeth commented:
16.04 has a too old version of `psutil`. You need to install `psutil` from another source than than the official repositories. This is literally mentioned in this very thread and clearly mentioned under the module requirements.

I also remind you that @azaghal did recommend you use `#ansible` IRC or at the very least use a dedicated issue thread.

gundalow commented:
@jlacroix82 please raise a new issue. 
This thread is finished as the original PR has been merged, thanks again for all the reviews on that.

siwyd commented:
Pointless thumbs up here considering it's already been merged, but I am using this to control Unity power settings + configuring Vino VNC settings and it works perfectly so far. Thanks for this!

gundalow commented:
@simonwydooghe Nice to know that people are finding the module useful and it's working for them :)
